docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,293 @@
+apparent in the expression.  In particular, field inference cannot change the
+outcome of initializer inference.
+
+
+A meta-goal is to keep this inference relatively lightweight and efficient to
+implement.
+
+
+In all cases, if initializer inference fails, it is an error and produces no
+result.  Implementations are free to treat this as a result of dynamic after
+emitting the error for the purposes of emitting subsequent error messages, but
+the behavior of programs for which initializer inference fails is unspecified.
+
+#### Immediately evident expressions
+
+We define a set of expressions with immediately-evident types (called
+immediately-evident expressions, abbreviated IE) as follows.  Immediately
+evident expressions have an inferred type as specified below.  Any expression
+which is not an IE has no inferred type.  For every immediately evident
+expression, we also define the set of variable dependencies upon which its
+inference relies.
+
+
+* null, boolean, numeric, string, and symbol literals have their corresponding
+   type.
+   * No inference dependencies
+* A list literal or a map literal with explicit type arguments has the type
+   `List<T>` or `Map<K, V>` respectively, where `<T>` or `<K, V>` are the
+   provided type arguments.
+   * No inference dependencies
+* A list literal with no type arguments, and for which all of the elements are
+   IEs, has the type `List<T>` where `T` is the least upper bound of the
+   inferred types of the elements.
+   * If the elements do not all have an inferred type, it is an error
+   * The inference dependencies of the list literal are the collected
+      dependencies of the elements.
+* A map literal with no type arguments, and for which all of the keys and
+   values are IEs, has the type `Map<K, V>` where `K` is the least upper bound of
+   the inferred types of the provided keys, and `V` is the least upper bound of
+   the provided values.
+   * If the keys and values do not all have an inferred type, it is an error.
+   * The inference dependencies of the map literal are the collected
+      dependencies of the keys and values.
+* A type literal has type `Type`
+   * No inference dependencies
+* A function literal with an expression body for which all parameter types are
+   specified and for which the return expression is an IE and has an inferred
+   type, has the corresponding function type.
+   * If the return expression has no inferred type it is an error.
+   * The inference dependencies are the inference dependencies of the return
+      expression.
+* An instance creation expression with no omitted type arguments has the
+   obvious type.
+   * No inference dependencies
+* An `as` expression has the cast to type
+   * No inference dependencies
+* A simple or qualified identifier referring to a top level function, getter,
+   or a static class getter or method, has the inferred type of the identifier.
```

Comment:
Does this include top level variables and static fields too? If so, might be good to call that even though technically they are synthesized getters.